### PR TITLE
[release-4.15] OCPBUGS-37532: fix leap file corner cases

### PR DESCRIPTION
**Pull request by vitus133**
This commit fixes various leap file corner cases and adds missing functionality:
Rehash the leapfile after setting the expiration time to 2036
Fix corner cases of leap event appearing in a different ways in the leap indication in the past or in the future + add unit tests
General cleanup and refactor for better testability

This is a manual cherry-pick of https://github.com/openshift/linuxptp-daemon/pull/324
/cc @aneeshkp @jzding @josephdrichard @nishant-parekh

### CODE REVIEW

File: pkg/leap/leap-file.go [hunk 1]
```diff
@@ -60,20 +63,22 @@ type LeapFile struct {
...
-	err := lm.PopulateLeapData()
+	err := lm.populateLeapData()
 	if err != nil {
 		return nil, err
 	}
 	return lm, nil
 }
 
-func ParseLeapFile(b []byte) (*LeapFile, error) {
+func parseLeapFile(b []byte) (*LeapFile, error) {
```
Comment (josephdrichard) [inline pkg/leap/leap-file.go:81]: Can you do the same validation for the prefixes that start with #?  At least validate that len(fields)>=2 so this can't segfault.

File: go.mod [hunk 2]
```diff
@@ -15,6 +15,7 @@ require (
 	github.com/stratoberry/go-gpsd v1.1.0
 	github.com/stretchr/testify v1.8.0
 	golang.org/x/sync v0.1.0
+	k8s.io/api v0.25.4
```
Comment (josephdrichard) [inline go.mod:18]: Is this change necessary?

Comment (vitus133) [reply go.mod:18]: Looks like client-go requires that:
k8s.io/client-go@v0.25.4 k8s.io/api@v0.25.4
This appeared after I added unit tests that use fake client

File: pkg/leap/leap-file.go (hunk 1 above)
Comment (vitus133) [reply pkg/leap/leap-file.go:82]: Done

File: pkg/leap/leap-file.go [hunk 3]
```diff
@@ -91,7 +96,10 @@ func ParseLeapFile(b []byte) (*LeapFile, error) {
 			}
 			sec, err := strconv.ParseInt(fields[1], 10, 8)
 			if err != nil {
-				return nil, fmt.Errorf("failed to parse Leap seconds %s value from file: %s, %v", fields[1], defaultLeapFileName, err)
+				return nil, fmt.Errorf("failed to parse Leap seconds %s value: %s", fields[1], err)
+			}
+			if _, err := strconv.ParseInt(fields[0], 10, 64); err != nil {
```
[the remark this answers is not in the document]
Comment (vitus133) [reply pkg/leap/leap-file.go]: Done

**Review by josephdrichard — APPROVED** [review]
/lgtm
/approved

**Comment by aneeshkp** [comment]
/re-title [release-4.15] OCPBUGS-37532: fix leap file corner cases
